### PR TITLE
Shortcuts, Auto-Return, and Grammar Updates

### PR DESCRIPTION
**Pull request by jdconrad**
* Added both shortcuts overloading the dot (.) operator and the brace operator ([]).  Dot operator can accept strings for maps and ints for lists, while the brace operator can accept expressions converted to objects for maps and expressions converted to int for lists.
* Added auto-return that will return the last value on the stack in the final statement (or final branches).
* Split the parser and the lexer to support modes in the lexer for the overloaded (.) operator.
* Fixed a bug in the grammar related to issues parsing types.

### CODE REVIEW

File: src/main/java/org/elasticsearch/plan/a/Writer.java [hunk 1]
```diff
@@ -1656,11 +1783,18 @@ private void writeLoadStoreExternal(final ParserRuleContext source) {
...
+        if (!length && !variable && !field && !array && !name && !shortcut) {
             throw new IllegalStateException(error(source) + "Target not found for load/store.");
         }
 
+        final boolean maplist = shortcut && (boolean)((Object[])sourceenmd.target)[2];
+        final Object constant = shortcut ? ((Object[])sourceenmd.target)[3] : null;
+
+        final boolean x1 = field || name || shortcut && !maplist;
+        final boolean x2 = array || shortcut && maplist;
+
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/plan/a/Writer.java]: can we add parens to these rather complex boolean expressions? makes it more readable for folks who have not memorized precedence :)

File: src/test/resources/rest-api-spec/test/plan_a/20_scriptfield.yaml [hunk 2]
```diff
@@ -19,7 +19,7 @@ setup:
                 script_fields:
                     bar:
                         script: 
-                            inline: "return input.get(\"doc\").get(\"foo\").get(0) + input.get(\"x\");"
+                            inline: "input.doc.foo.0 + input.x;"
                             lang: plan-a
```
Comment (rmuir) [inline src/test/resources/rest-api-spec/test/plan_a/20_scriptfield.yaml]: yes!

**Comment by rmuir** [comment]
+1, i added a note (just a general observation). It might help make code more approachable, to people who don't work with parsers everyday!

File: src/main/java/org/elasticsearch/plan/a/Writer.java (hunk 1 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/plan/a/Writer.java]: Done.

**Comment by jdconrad** [comment]
This was checked in via master.  Thanks for the review!

**Pull request closed without merging**